### ISSUE TITLE
ec2_group module indicates boto3 not installed, even though it is installed

### ISSUE THREAD

**Issue by edesz** [ansible/ansible#33052]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
`ec2_group`

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = /home/username/ansible-proj/ansible.cfg
  configured module search path = [u'/home/username/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/username/ansible-proj/.venv/local/lib/python2.7/site-packages/ansible
  executable location = /home/username/ansible-proj/.venv/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```

##### CONFIGURATION
```
DEFAULT_VAULT_PASSWORD_FILE(/home/username/ansible-proj/ansible.cfg) = /home/username/vault_pass.txt
(END)
```

##### OS / ENVIRONMENT
OS running Ansible from: Ubuntu 17.10

##### SUMMARY
I have the following basic Ansible `ec2_group` module-based task to create an AWS ec2 group. The `ec2_group` module has the requirement of `boto` and `boto3`, both of which I have installed. When I run the task below, it indicates that the module requires `boto3`. However, `boto3` is already installed.

##### STEPS TO REPRODUCE
Here is the task, based on the Ansible docs [examples](http://docs.ansible.com/ansible/latest/ec2_group_module.html#examples):

```
- hosts: localhost
  connection: local

  vars:
    region: "us-east-2"
    sg_name: "redhat-ec2"
    sg_desc: "Security group for Ansible RedHat"
    sg_rules:
      - proto: tcp
        from_port: 80
        to_port: 80
        cidr_ip: 0.0.0.0/0
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: 0.0.0.0/0

  tasks:
    - name: create security group for redhat ec2 instance(s)
      ec2_group:
      region: "{{ region }}"
      name: "{{ sg_name }}"
      description: "{{ sg_desc }}"
      rules: "{{ sg_rules|list }}"
      rules_egress:
        - proto: all
          cidr_ip: 0.0.0.0/0
```

Install relevant packages with:

```
mkdir -p ~/ansible-proj
cd ansible-proj
virtualenv .
source .venv/bin/activate
pip install ansible
pip install awscli boto
pip install boto3
```

Run with `ansible-playbook -i /.hosts mytest.yml`. Here is `hosts`:

```
[localhost]
my_machine ansible_host=127.0.0.1 ansible_become_pass=<mypasswordhere>
```

##### EXPECTED RESULTS
The above task should create an AWS `ec2_group`

##### ACTUAL RESULTS
WITHOUT verbosity

When I run the above Ansible task, I get a message indicating `boto3` is needed - however `boto3` is already installed:

```
PLAY [Initial configuration of AWS CLI on Ansible controller] ****************************

TASK [Gathering Facts] *******************************************************************
ok: [my_machine]

TASK [ansible-test : create security group] ***********
fatal: [my_machine]: FAILED! => {"changed": false, "failed": true, "msg": "boto3 required for this module"}
```

WITH verbosity -vvvv

```
fatal: [my_machine]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "description": "Security group for Ansible RedHat", 
            "ec2_url": null, 
            "group_id": null, 
            "name": "redhat-ec2", 
            "profile": null, 
            "purge_rules": true, 
            "purge_rules_egress": true, 
            "purge_tags": true, 
            "region": "us-east-2", 
            "rules": [
                {
                    "cidr_ip": "0.0.0.0/0", 
                    "from_port": 80, 
                    "proto": "tcp", 
                    "to_port": 80
                }, 
                {
                    "cidr_ip": "0.0.0.0/0", 
                    "from_port": 22, 
                    "proto": "tcp", 
                    "to_port": 22
                }
            ], 
            "rules_egress": [
                {
                    "cidr_ip": "0.0.0.0/0", 
                    "proto": "all"
                }
            ], 
            "security_token": null, 
            "state": "present", 
            "tags": null, 
            "validate_certs": true, 
            "vpc_id": null
        }
    }, 
    "msg": "boto3 required for this module"
}
```

##### OTHER

Here is my `pip freeze` output:

```
ansible==2.4.1.0
.
.
.
boto==2.48.0
boto3==1.4.7
botocore==1.7.47
.
.
.
paramiko==2.4.0
.
.
six==1.11.0
```

The main packages I have installed through `pip` are:

```
ansible
pywinrm
boto
awscli
boto3
sshpass
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/amazon/ec2_group.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_group.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @adq @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by s-hertel** [comment]
@edesz I can't reproduce this, though I have run into something similar before. In your virtualenv can you manually try 
```
>>> import boto3
>>> boto3.client('ec2')
<botocore.client.EC2 object at 0x102c1c5d0>
>>> import botocore.exceptions
```
If that doesn't work, the wrong library may be imported so you might want to check the `__path__`.

needs_info

**Comment by willthames** [comment]
The problem could also be that ansible isn't using the virtualenv's python correctly.

You ran the task under `-vvvv` but missed the most important part of the output, which is what python interpreter ran the module. Can you paste the full output of the task?

**Comment by edesz** [comment]
@s-hertel Ok, I ran those 3 Python commands and I got the same output as you did. There were no issues with executing them. It seems like it is correctly imported.

EDIT: I ran all the commands inside the `virtualenv`.

**Comment by edesz** [comment]
@willthames Do you mean that you want to know is the system-wide one executing the module, or the one in the `venv`?

Here is the full output of the task:

```
ansible-playbook 2.4.1.0
  config file = /home/username/ansible-proj/ansible.cfg
  configured module search path = [u'/home/username/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/username/ansible-proj/.venv/local/lib/python2.7/site-packages/ansible
  executable location = /home/username/ansible-proj/.venv/bin/ansible-playbook
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
Using /home/username/ansible-proj/ansible.cfg as config file
setting up inventory plugins
Parsed /home/username/ansible-proj/hosts inventory source with ini plugin
Loading callback plugin default of type stdout, v2.0 from /home/username/ansible-proj/.venv/local/lib/python2.7/site-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: mytest.yml *************************************************
1 plays in mytest.yml

PLAY [Initial configuration of AWS CLI on Ansible controller] ******************
Trying secret FileVaultSecret(filename='/home/username/vault_pass.txt') for vault_id=default

TASK [Gathering Facts] *********************************************************
Using module file /home/username/ansible-proj/.venv/local/lib/python2.7/site-packages/ansible/modules/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: username
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/username/.ansible/tmp/ansible-tmp-1511322194.24-98788698105341 `" && echo ansible-tmp-1511322194.24-98788698105341="` echo /home/username/.ansible/tmp/ansible-tmp-1511322194.24-98788698105341 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmp02EJI6 TO /home/username/.ansible/tmp/ansible-tmp-1511322194.24-98788698105341/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/username/.ansible/tmp/ansible-tmp-1511322194.24-98788698105341/ /home/username/.ansible/tmp/ansible-tmp-1511322194.24-98788698105341/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/username/.ansible/tmp/ansible-tmp-1511322194.24-98788698105341/setup.py; rm -rf "/home/username/.ansible/tmp/ansible-tmp-1511322194.24-98788698105341/" > /dev/null 2>&1 && sleep 0'
ok: [my_machine]
META: ran handlers

TASK [ansible-test : create security group] ***
task path: /home/username/ansible-proj/mytest.yml:21
Using module file /home/username/ansible-proj/.venv/local/lib/python2.7/site-packages/ansible/modules/cloud/amazon/ec2_group.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: username
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179 `" && echo ansible-tmp-1511322197.34-225924969427179="` echo /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpDe9FFH TO /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ec2_group.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ec2_group.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ec2_group.py; rm -rf "/home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/" > /dev/null 2>&1 && sleep 0'
fatal: [my_machine]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "description": "Security group for Ansible RedHat", 
            "ec2_url": null, 
            "group_id": null, 
            "name": "redhat-ec2", 
            "profile": null, 
            "purge_rules": true, 
            "purge_rules_egress": true, 
            "purge_tags": true, 
            "region": "us-east-2", 
            "rules": [
                {
                    "cidr_ip": "0.0.0.0/0", 
                    "from_port": 80, 
                    "proto": "tcp", 
                    "to_port": 80
                }, 
                {
                    "cidr_ip": "0.0.0.0/0", 
                    "from_port": 22, 
                    "proto": "tcp", 
                    "to_port": 22
                }
            ], 
            "rules_egress": [
                {
                    "cidr_ip": "0.0.0.0/0", 
                    "proto": "all"
                }
            ], 
            "security_token": null, 
            "state": "present", 
            "tags": null, 
            "validate_certs": true, 
            "vpc_id": null
        }
    }, 
    "msg": "boto3 required for this module"
}
	to retry, use: --limit @/home/username/ansible-proj/mytest.retry

PLAY RECAP *********************************************************************
my_machine           : ok=3    changed=0    unreachable=0    failed=1   
```

**Comment by willthames** [comment]
Yep, as I thought:

```

TASK [ansible-test : create security group] ***
task path: /home/username/ansible-proj/mytest.yml:21
Using module file /home/username/ansible-proj/.venv/local/lib/python2.7/site-packages/ansible/modules/cloud/amazon/ec2_group.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: username
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179 `" && echo ansible-tmp-1511322197.34-225924969427179="` echo /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpDe9FFH TO /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ec2_group.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ec2_group.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/ec2_group.py; rm -rf "/home/username/.ansible/tmp/ansible-tmp-1511322197.34-225924969427179/" > /dev/null 2>&1 && sleep 0'
```

In the first line of the above, you can see that the expected python is used for running stuff on your local machine. In the last line, however, when running the module, /usr/bin/python is used. This is because this defaults to `ansible_python_interpreter`

One problem is that although `ansible_playbook_python` exists, you can't set `ansible_python_interpreter` at runtime

If you use `delegate_to: localhost` in the task, rather than `connection: local`, this problem goes away (which is a giant chore because you can't delegate at a playbook level, only at a task level)

cc'ing @bcoca, @abadger, @jimi-c as this has bitten me so many times I'm just annoyed with it now

**Comment by edesz** [comment]
Yup, I see it!

**Comment by dagwieers** [comment]
@willthames Allowing `delegate_to` at the playbook level is something I would like to see as well.

**Comment by edesz** [comment]
I seem to have overcome this problem by specifying `ansible_python_interpreter` as @willthames mentioned above and also as shown in the example [here](https://github.com/ansible/ansible/issues/21468#issuecomment-317189249).

I tested each of the following approaches (separately):
1. `hosts` file:
```
[localhost]
my_machine ansible_host=127.0.0.1 ansible_become_pass=<mypasswordhere>
      ansible_python_interpreter=/home/username/ansible-proj/.venv/bin/python
```

2. Create `group_vars/all/all.yml` file (and make no changes to `hosts`):
```
---
ansible_python_interpreter=/home/username/ansible-proj/.venv/bin/python
```

Both of these approaches seem to have worked. I kept `connection: local` in my main playbook `mytest.yml` - the tasks in `mytest.yml` were also unchanged. As before, I ran with `ansible-playbook -i /.hosts mytest.yml`.

I've just run the playbook 3 times since making this change but it seems to be working so far.

**Comment by willthames** [comment]
While I'm glad you've found a solution for your use case, this is still indicative of a wider problem (you can't share your inventory with other people as they'll likely have a completely different requirement for ansible_python_interpreter)

**Comment by edesz** [comment]
Yup, you're right about that. It makes complete sense to me from a portability point of view.

**Comment by willthames** [comment]
Closing this one now.

Thanks to #33698, you can now set `ansible_python_interpreter: {{ ansible_playbook_python }}` in e.g. host_vars/localhost.yml

**Comment by emre141** [comment]
Hi,

I have same problem just using aws_ssm lookup

name : Get SSM Parameter Store RAbbitMQ Username
debug: msg="{{ lookup('aws_ssm', 'rabbitmq_username', region='eu-central-1' ) }}"
I am using ansible role in my autoscaling group so there is a cloud-init file to run ansible-playbook as below

ansible-playbook /etc/ansible/roles/rabbitmq/tests/test.yml --connection=local

So new instance in ASG initializating execute above command then there are some task to use yum module and pip module before using aws_ssm module for install all requirement but i have got below erron on aws_ssm task

TASK [rabbitmq : Install python-pip and other tools to install python packages/ansible] ***
[DEPRECATION WARNING]: Invoking "yum" only once while using a loop via
squash_actions is deprecated. Instead of using a loop to supply multiple items
and specifying name: {{ item }}, please use name: [u'python-pip', u'yum- plugin-versionlock', u'awslogs', u'python2-boto3'] and remove the loop. This
feature will be removed in version 2.11. Deprecation warnings can be disabled
by setting deprecation_warnings=False in ansible.cfg.
changed: [localhost] => (item=[u'python-pip', u'yum-plugin-versionlock', u'awslogs', u'python2-boto3'])

TASK [rabbitmq : Install Boto & Ansible to run Ansible when instance comes up] ***
[DEPRECATION WARNING]: Invoking "pip" only once while using a loop via
squash_actions is deprecated. Instead of using a loop to supply multiple items
and specifying name: {{ item }}, please use name: [u'pip', u'jinja2', u'boto', u'boto3', u'botocore', u'awscli', u'cryptography'] and remove the
loop. This feature will be removed in version 2.11. Deprecation warnings can be
disabled by setting deprecation_warnings=False in ansible.cfg.
changed: [localhost -> localhost] => (item=[u'pip', u'jinja2', u'boto', u'boto3', u'botocore', u'awscli', u'cryptography'])

TASK [rabbitmq : command] ******************************************************
changed: [localhost]

TASK [rabbitmq : command] ******************************************************
changed: [localhost]

TASK [rabbitmq : Get SSM Parameter Store RAbbitMQ Username] ********************
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'aws_ssm'. Error was a <class 'ansible.errors.AnsibleError'>, original message: botocore and boto3 are required for aws_ssm lookup."}
to retry, use: --limit @/etc/ansible/roles/rabbitmq/tests/test.retry

It works when i run my role while ec2 instance running.After that the instance wake up when i execute role again manually via connected to the instance it works and complete successfull but this problemlem only occured new instance initialization

[root@ip-172-16-121-52 ec2-user]# ansible --version
ansible 2.7.0
config file = /etc/ansible/ansible.cfg
configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
ansible python module location = /usr/lib/python2.7/site-packages/ansible
executable location = /bin/ansible
python version = 2.7.14 (default, Jul 26 2018, 19:59:38) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
[root@ip-172-16-121-52 ec2-user]# pip --version
pip 18.1 from /usr/lib/python2.7/site-packages/pip (python 2.7)
[root@ip-172-16-121-52 ec2-user]# pip list | grep boto
boto 2.49.0
boto3 1.9.24
botocore 1.12.24

 👍 1
Sign up for free

**Comment by willthames** [comment]
@emre141 can you raise this as a new issue, this one is closed.

In the new issue, please run the playbook with `-vvv` and paste the results in enclosed in triple backticks so that it's a little easier to read.

I also suggest adding a `python_requirements_facts` task to the playbook after the pip install to ensure that it's installed correctly from ansible's point of view

```
python_requirements_facts:
  - botocore
  - boto3
  - boto
```
